internal/assets/token_pool_created.go

```diff
@@ -17,69 +17,78 @@
 package assets
 
 import (
-	"context"
-
 	"github.com/hyperledger/firefly/internal/log"
 	"github.com/hyperledger/firefly/pkg/database"
 	"github.com/hyperledger/firefly/pkg/fftypes"
 	"github.com/hyperledger/firefly/pkg/tokens"
 )
 
-func (am *assetManager) persistTokenPoolTransaction(ctx context.Context, pool *fftypes.TokenPool, signingIdentity string, protocolTxID string, additionalInfo fftypes.JSONObject) (valid bool, err error) {
-	if pool.ID == nil || pool.TX.ID == nil {
-		log.L(ctx).Errorf("Invalid token pool '%s'. Missing ID (%v) or transaction ID (%v)", pool.ID, pool.ID, pool.TX.ID)
-		return false, nil // this is not retryable
+func (am *assetManager) TokenPoolCreated(tk tokens.Plugin, tokenType fftypes.TokenType, tx *fftypes.UUID, protocolID, signingIdentity, protocolTxID string, additionalInfo fftypes.JSONObject) error {
+	// Find a matching operation within this transaction
+	fb := database.OperationQueryFactory.NewFilter(am.ctx)
+	filter := fb.And(
+		fb.Eq("tx", tx),
+		fb.Eq("type", fftypes.OpTypeTokensCreatePool),
+	)
+	operations, _, err := am.database.GetOperations(am.ctx, filter)
+	if err != nil || len(operations) == 0 {
+		log.L(am.ctx).Debugf("Token pool transaction '%s' ignored, as it was not submitted by this node", tx)
+		return nil
+	}
+
+	pool := &fftypes.TokenPoolAnnouncement{
+		TokenPool: fftypes.TokenPool{
+			Type:       tokenType,
+			ProtocolID: protocolID,
+			Author:     signingIdentity,
+		},
+		ProtocolTxID: protocolTxID,
 	}
-	return am.txhelper.PersistTransaction(ctx, &fftypes.Transaction{
-		ID: pool.TX.ID,
+	err = retrieveTokenOpInputs(am.ctx, operations[0], &pool.TokenPool)
+	if err != nil {
+		log.L(am.ctx).Errorf("Error retrieving pool info from transaction '%s' (%s) - ignoring: %v", tx, err, operations[0].Input)
+		return nil
+	}
+
+	// Update the transaction with the info received (but leave transaction as "pending")
+	transaction := &fftypes.Transaction{
+		ID:     tx,
+		Status: fftypes.OpStatusPending,
 		Subject: fftypes.TransactionSubject{
 			Namespace: pool.Namespace,
-			Type:      pool.TX.Type,
+			Type:      fftypes.TransactionTypeTokenPool,
 			Signer:    signingIdentity,
 			Reference: pool.ID,
 		},
 		ProtocolID: protocolTxID,
 		Info:       additionalInfo,
-	})
-}
-
-func (am *assetManager) persistTokenPool(ctx context.Context, pool *fftypes.TokenPool) (valid bool, err error) {
-	l := log.L(ctx)
-	if err := fftypes.ValidateFFNameField(ctx, pool.Name, "name"); err != nil {
-		l.Errorf("Invalid token pool '%s' - invalid name '%s': %a", pool.ID, pool.Name, err)
-		return false, nil // This is not retryable
 	}
-	err = am.database.UpsertTokenPool(ctx, pool)
-	if err != nil {
-		if err == database.IDMismatch {
-			log.L(ctx).Errorf("Invalid token pool '%s'. ID mismatch with existing record", pool.ID)
-			return false, nil // This is not retryable
+	pool.TX.ID = transaction.ID
+	pool.TX.Type = transaction.Subject.Type
+
+	// Add a new operation for the announcement
```

Comment:
Wasn't sure if we want an operation for this leg of the transaction or not (ie tracking the state of the broadcast announcement). I went ahead and added one, but I'm not positive it's adding value.